internal/future/statestore/redis.go

```diff
@@ -94,29 +96,208 @@ func redisCloser(closer func() error) {
+	redisConn, err := rb.redisPool.GetContext(ctx)
+	if err != nil {
+		redisLogger.WithFields(logrus.Fields{
+			"error": err.Error(),
+		}).Error("failed to connect to redis")
+		return nil, err
+	}
+	defer redisConn.Close()
+
+	value, err := redis.Bytes(redisConn.Do("GET", id))
+	if err != nil {
+		redisLogger.WithFields(logrus.Fields{
+			"key":   id,
+			"cmd":   "GET",
+			"error": err.Error(),
+		}).Error("failed to get the ticket from state storage")
+		return nil, err
+	}
+
+	if value == nil {
+		msg := fmt.Sprintf("Ticket id:%s not found", id)
+		redisLogger.WithFields(logrus.Fields{
+			"key": id,
+			"cmd": "GET",
+		}).Error(msg)
+		return nil, status.Error(codes.NotFound, msg)
+	}
+
+	ticket := &pb.Ticket{}
+	err = proto.Unmarshal(value, ticket)
+	if err != nil {
+		redisLogger.WithFields(logrus.Fields{
+			"key":   id,
+			"error": err.Error(),
+		}).Error("failed to unmarshal the ticket proto")
+		return nil, err
+	}
+
+	return ticket, nil
 }
 
-// DeleteTicket removes the Ticket with the specified id from state storage. This method succeeds if the Ticket does not exist.
-func (rb *redisBackend) DeleteTicket(context.Context, string) error {
-	return status.Error(codes.Unimplemented, "not implemented")
+// DeleteTicket removes the Ticket with the specified id from state storage.
+func (rb *redisBackend) DeleteTicket(ctx context.Context, id string) error {
+	redisConn, err := rb.redisPool.GetContext(ctx)
+	if err != nil {
+		redisLogger.WithFields(logrus.Fields{
+			"error": err.Error(),
+		}).Error("failed to connect to redis")
+		return err
+	}
+	defer redisConn.Close()
+
+	_, err = redisConn.Do("DEL", id)
+	if err != nil {
+		redisLogger.WithFields(logrus.Fields{
+			"key":   id,
+			"cmd":   "DEL",
+			"error": err.Error(),
+		}).Error("failed to delete the ticket from state storage")
+		return err
+	}
+
+	return nil
 }
 
 // IndexTicket indexes the Ticket id for the specified fields.
-func (rb *redisBackend) IndexTicket(context.Context, pb.Ticket, []string) error {
-	return status.Error(codes.Unimplemented, "not implemented")
+func (rb *redisBackend) IndexTicket(ctx context.Context, ticket pb.Ticket, indices []string) error {
+	redisConn, err := rb.redisPool.GetContext(ctx)
+	if err != nil {
+		redisLogger.WithFields(logrus.Fields{
+			"error": err.Error(),
+		}).Error("failed to connect to redis")
+		return err
+	}
+	defer redisConn.Close()
+
+	redisConn.Send("MULTI")
+	// Loop through all attributes we want to index.
+	for _, attribute := range indices {
+		// TODO: Currently, Open Match supports specifying attributes as JSON properties on Tickets.
+		// Alternatives for populating this information in proto fields are being considered.
+		// Also, we need to add Ticket creation time to either the ticket or index.
+		// Meta characters bein specified in JSON property keys is currently not supported.
+		v := gjson.Get(ticket.Properties, attribute)
```

Comment:
Just use the `import "encoding/json"` package for this.

Reply:
Given that we don't know the structure of the Properties, gjson just fetches the value for the given attribute. Since we are considering moving away from JSON, i ll leave it as such. If we decide to stick to JSON, then we should at-least determine the structure and use encoding/json to unmarshal.

Reply:
I did some tests on the gjson package and it turns out gjson.Get does not search for the attribute recursively. For example.

```
gjson.Get("{\"hey\": \"false\", \"hi\": {\"hello\": \"true\"}", "hello")          // returns nothing
gjson.Get("{\"hey\": \"false\", \"hi\": {\"hello\": \"true\"}", "hi.hello")       // returns true
gjson.Get("{\"hey\": \"false\", \"hi\": {\"hello\": \"true\"}", "hey")            // returns false
```

so `gjson.Get` here basically is equivalent to 
```
m := make(map[string]interface{})
err := json.Unmarshal(data, &m)
if err != nil {
    log.Fatal(err)
}
fmt.Println(m[attribute])
```

+1 for using Any/determine JSON structure and use encoding/json